
<html>

    <head>
        <title>Joseph Crispell | Blog - Consensus</title>
        <link href="../../default.css" rel="stylesheet" type="text/css">
        
        <!-- Global site tag (gtag.js) - Google Analytics -->
        <script async src="https://www.googletagmanager.com/gtag/js?id=UA-109829558-1"></script>
        <script>
            window.dataLayer = window.dataLayer || [];
            function gtag(){dataLayer.push(arguments);}
            gtag('js', new Date());
            gtag('config', 'UA-109829558-1');
        </script>
    </head>

    <body>
        
        <!-- Insert Sidebar box -->
        <div id="sidebar" style="top: 43.5%">
            
            <pageLink><a href="../../index.html">Home</a></pageLink>
            <pageLink><a href="../../Blog.html">Blog</a></pageLink>
            <pageLink><a href="../../About.html">About</a></pageLink>
            <pageLink><a href="../../Outputs.html">Outputs</a></pageLink>
        </div>
        
        <!-- Insert Title -->
        <div id="title">
            <img src="../../BlogThumbnail.png" width="40%">
                
        </div>
        
        <!-- Insert Text Body -->
        <div id="mainText">
            
            <br><br><br>
            
            <h1>Building a consensus sequence&emsp;11-12-17</h1>
            
            <p>
                Off to a meeting in Birmingham to hear about all the <a href="https://www.bbsrc.ac.uk/">BBSRC</a> funded research bovine tuberculosis research. A super early flight and I thought I'd get
                this post written whilst I am travelling.<br><br>
                
                I've been meaning to note down how to build a consensus sequence from multiple unaligned sequences. Once sequences are aligned to one another the consensus sequence is the sequence that
                selects the most common nucleotide (A, C, G, or T) present at each position.<br><br>
                
                Before a consensus sequence can be built, the sequences must be aligned. Thankfully there are plenty of alignment tools out there. Here, I use <a href="http://www.clustal.org/">Clustal</a>.
                <a href="http://www.clustal.org/">Clustal</a> can be used in <a href="https://www.r-project.org/">R</a> but first you'll need to install it (download
                <a href="http://www.clustal.org/download/current/">here</a>) then add it into your path variable (on a Windows machine, this is done automatically on Ubuntu). Or just type the following in the Ubuntu terminal:<br><br>
			</p>
			
			<div id="codeBlock">
                sudo apt update<br>
				sudo apt install clustal
            </div>
			
			<p>
                The path variable is an interesting thing - whenever you call a program by its name (here <a href="https://www.r-project.org/">R</a> is calling <a href="http://www.clustal.org/">Clustal</a>),
                the computer will search for that program using the path variable. A file path is the ordered list of computer folder names, separated by a slash ("/" on mac and unix or "\" on windows),
                where a file can be found. For example, "test.txt" can be found using the following path: "/Users/josephcrisp1/Desktop/test.txt". The path variable is list, separated by ":", of the paths to
                different programs on your computer.<br><br>
                
                Installing and adding <a href="http://www.clustal.org/">Clustal</a> to the path is the hardest part, once that's' done use the following <a href="https://www.r-project.org/">R</a> code to
                align FASTA sequences:
            </p>
            
            <div id="codeBlock">
                # Load the ape library - used for handling sequencing data in R<br>
                library(ape)<br><br>
                
                # Read in the FASTA sequences<br>
                sequences &lt;- read.FASTA("pathToFile")<br><br>
                
                # Run Clustal multiple sequence alignment<br>
                alignment &lt;- clustal(sequences)
            </div>
            
            <p>
                Now I built an <a href="https://www.r-project.org/">R</a> function that can build a consensus sequence aligned sequences:
            </p>
            
            <div id="codeBlock">
                getConsensus &lt;- function(sequences){<br><br>
                    
                &emsp;&emsp;# Convert between ways of storing sequences<br>
                &emsp;&emsp;sequences &lt;- as.alignment(sequences)<br><br>
                    
                &emsp;&emsp;# Initialise vector to store the consensus<br>
                &emsp;&emsp;consensus &lt;- c()<br><br>
				
				&emsp;&emsp;# Get the sequences and split them into their nucleotides<br>
				&emsp;&emsp;nucleotides &lt;- list()<br>
				&emsp;&emsp;for(i in 1:sequences$nb){<br>
				&emsp;&emsp;&emsp;&emsp;nucleotides[[sequences$nam[i]]] &lt;- strsplit(sequences$seq[i], split="")[[1]]<br>
				&emsp;&emsp;}<br><br>
                    
                &emsp;&emsp;# Examine each position in the alignment<br>
                &emsp;&emsp;for(position in 1:length(nucleotides[[1]])){<br><br>
                    
                &emsp;&emsp;&emsp;&emsp;# Initialise a vector to store the alleles at the current position<br>
                &emsp;&emsp;&emsp;&emsp;alleles &lt;- c()<br><br>
                    
                &emsp;&emsp;&emsp;&emsp;# Examine the current position in each sequence - store each allele<br>
                &emsp;&emsp;&emsp;&emsp;for(i in 1:sequences$nb){<br>
                &emsp;&emsp;&emsp;&emsp;&emsp;&emsp;alleles[i] &lt;- nucleotides[[i]][position]<br>
                &emsp;&emsp;&emsp;&emsp;}<br><br>
                    
                &emsp;&emsp;&emsp;&emsp;# Get the number of each allele present<br>
				&emsp;&emsp;&emsp;&emsp;alleleCounts &lt;- table(alleles)<br><br>
      
				&emsp;&emsp;&emsp;&emsp;# Find the most frequent allele and assign to consensus<br>
				&emsp;&emsp;&emsp;&emsp;maxAllele &lt;- names(which(alleleCounts == max(alleleCounts)))<br>
				&emsp;&emsp;&emsp;&emsp;if(length(maxAllele) == 1){<br>
				&emsp;&emsp;&emsp;&emsp;&emsp;consensus[position] &lt;- maxAllele<br>
				&emsp;&emsp;&emsp;&emsp;}else{<br>
				&emsp;&emsp;&emsp;&emsp;&emsp;consensus[position] &lt;- "n"<br>
				&emsp;&emsp;&emsp;&emsp;}<br>
                &emsp;&emsp;}<br><br>
				
				&emsp;&emsp;# Attach the consensus to the alignment<br>
				&emsp;&emsp;sequences$seq[sequences$nb + 1] &lt;- paste(consensus, collapse="")<br>
				&emsp;&emsp;sequences$nb &lt;- sequences$nb + 1<br>
				&emsp;&emsp;sequences$nam &lt;- c(sequences$nam, "consensus")<br><br>
				
				&emsp;&emsp;# Convert between ways of storing sequences<br>
				&emsp;&emsp;sequences &lt;- as.DNAbin.alignment(sequences)<br><br>
                    
                &emsp;&emsp;return(sequences)<br>
                }
            </div>
            
			<p>
            	A pretty awful looking function. Basically it first gets the alignment into a manageable format (a list of nucleotide sequences), and then records the most common allele found at each position and builds these into the consensus sequence.<br><br>
				
			</p>
			
			<div id="codeBlock">
                # Add the consensus sequence into the alignment<br>
				alignment &lt;- getConsensus(alignment)<br>
            </div>
			
			<p>				
				The <c>getConsensus()</c> function adds the consensus sequence onto the alignment. To save the alignment to file, use the following R code:
            </p>
			
			<div id="codeBlock">
                # Print out sequences with consensus<br>
                outputFile &lt;- paste(path, "sequencesWithConsensus.fasta", sep="")<br>
                write.FASTA(alignment, file=outputFile)
            </div>
			
            <p>
                So that's it! I hope some of the above code is useful, it can all be found in this <a href="https://github.com/JosephCrispell/GeneralTools/blob/master/GetConsensusFromClustalAlignment_28-09-17.R">script</a>.
            </p>
            
        </div>

    </body>

</html>
