
<html>

    <head>
        <title>Joseph Crispell | Blog - BLAST</title>
        <link href="../../default.css" rel="stylesheet" type="text/css">
    </head>

    <body>
        
        <!-- Insert Sidebar box -->
        <div id="sidebar" style="top: 43.5%">
            
            <pageLink><a href="../../index.html">Home</a></pageLink>
            <pageLink><a href="../../Blog.html">Blog</a></pageLink>
            <pageLink><a href="../../About.html">About</a></pageLink>
            <pageLink><a href="../../Outputs.html">Outputs</a></pageLink>
        </div>
        
        <!-- Insert Title -->
        <div id="title">
            <img src="../../BlogThumbnail.png" width="40%">
                
        </div>
        
        <!-- Insert Text Body -->
        <div id="mainText">
            
            <br><br><br>
            
            <h1>Creating and comparing sequences to a local BLAST database&emsp;03-08-17</h1>
            
            <p>
                <a href="https://blast.ncbi.nlm.nih.gov/Blast.cgi">BLAST</a> is a powerful tool that can be used to compare genetic sequences.
                In my research, I use <a href="https://blast.ncbi.nlm.nih.gov/Blast.cgi">BLAST</a> to check that my sequence data corresponds
                to my organism of interest and isn't badly contaminated. To do this, I align my sequence data to a reference genome (for my
                organism of interest) and use <a href="https://blast.ncbi.nlm.nih.gov/Blast.cgi">BLAST</a> to check the unaligned sequence data
                against online databases.<br><br>
                
                I have also used <a href="https://blast.ncbi.nlm.nih.gov/Blast.cgi">BLAST</a> to conduct local searches. Here, I have had a panel of sequences
                that I want to search for in some genetic data. To do this, I have built local <a href="https://blast.ncbi.nlm.nih.gov/Blast.cgi">BLAST</a>
                databases to compare my query sequences against.<br><br>
                
                In the UNIX environment <a href="https://blast.ncbi.nlm.nih.gov/Blast.cgi">BLAST</a> can easily be installed using the following code:
            </p>
            
            <div id="codeBlock">
                # Install BLAST<br>
                $ sudo apt-get install ncbi-blast+
            </div>
            
            <p>
                Once installed <a href="https://blast.ncbi.nlm.nih.gov/Blast.cgi">BLAST</a> can be executed from the UNIX terminal. You can either compare your
                sequences to all the nucleotide data available in the online databases, using the following code:
            </p>
            
            <div id="codeBlock">
                # Compare query sequences to online databases<br>
                $ blastn -query querySequences.fasta -out blastOutput.txt -db nr -remote
                
            </div>
            
            <p>
                Or you can create a local database on your computer containing the sequences that you are interested in comparing your query sequences against,
                using the following code:
            </p>
            
            <div id="codeBlock">
                # Make local database<br>
                $ makeblastdb -in databaseSequences.fasta -parse_seqids -dbtype nucl
                
            </div>
            
            <p>
                With your local database created, you can compare your query sequences against it using the following code:
            </p>
            
            <div id="codeBlock">
                # Compare sequences to local database<br>
                $ blastn -query querySequences.fasta -out blastOutput.txt -db databaseSequences -outfmt 0
            </div>
            
        </div>
    </body>
</html>
