
<html>

    <head>
        <title>Joseph Crispell | Blog - Multiple histograms</title>
        <link href="../../default.css" rel="stylesheet" type="text/css">

        <!-- Global site tag (gtag.js) - Google Analytics -->
        <script async src="https://www.googletagmanager.com/gtag/js?id=UA-109829558-1"></script>
        <script>
            window.dataLayer = window.dataLayer || [];
            function gtag(){dataLayer.push(arguments);}
            gtag('js', new Date());
            gtag('config', 'UA-109829558-1');
        </script>
    </head>

    <body>

        <!-- Insert Sidebar box -->
        <div id="sidebar" style="top: 43.5%">

            <pageLink><a href="../../index.html">Home</a></pageLink>
            <pageLink><a href="../../Blog.html">Blog</a></pageLink>
            <pageLink><a href="../../About.html">About</a></pageLink>
            <pageLink><a href="../../Outputs.html">Outputs</a></pageLink>
        </div>

        <!-- Insert Title -->
        <div id="title">
            <img src="../../BlogThumbnail.png" width="40%">

        </div>

        <!-- Insert Text Body -->
        <div id="mainText">

            <br><br><br>

            <h1>A quick look at my new function: plotFASTA()&emsp;27-10-19</h1>

            <p>
                I've added a new function to my <c>basicPlotteR</c> package, it's called <c>plotFASTA()</c>. This function is is a quick way to plot a nucleotide alignment.<br><br>

                To use it, first install my <c>basicPlotteR()</c> package:
            </p>

            <div id="codeBlock">

                # Install and load the devtools package<br>
                install.packages("devtools")<br><br>

                # Install and load the basicPlotteR package<br>
                devtools::install_github("JosephCrispell/basicPlotteR")<br>
                library(basicPlotteR)<br><br>
            </div>

            <p>
                Next, let's create a random nucleotide alignment. I've created a couple of functions that we'll use:
            </p>

            <div id="codeBlock">

                createNucleotideSequenceAlignment &lt;- function(nSequences, nSites, meanNSNPs){<br><br>
  
                &emsp;&emsp;# Build a random reference genome<br>
                &emsp;&emsp;reference &lt;- sample(c("A", "C", "G", "T"), replace=TRUE, size=nSites)<br><br>
  
                &emsp;&emsp;# Create a matrix to store the sequences<br>
                &emsp;&emsp;sequences &lt;- matrix(reference, byrow=TRUE, nrow=nSequences, ncol=nSites)<br>
                &emsp;&emsp;sequences[1, ] &lt;- reference<br>
                &emsp;&emsp;rownames(sequences)&lt;- paste0("Seq-", 1:nrow(sequences))<br><br>
  
                &emsp;&emsp;# Create mutations in the sequences<br>
                &emsp;&emsp;sequences &lt;- mutateSequences(sequences, meanNSNPs)<br><br>
  
                &emsp;&emsp;return(sequences)<br>
                }<br><br>

                mutateSequences &lt;- function(sequences, meanNSNPs){<br><br>
  
                &emsp;&emsp;for(row in 1:nrow(sequences)){<br><br>
    
                &emsp;&emsp;&emsp;&emsp;# Decide how many mutations will be in the current sequence<br>
                &emsp;&emsp;&emsp;&emsp;nMutations &lt;- rpois(1, lambda=meanNSNPs)<br><br>
    
                &emsp;&emsp;&emsp;&emsp;# Choose the position for the mutations<br>
                &emsp;&emsp;&emsp;&emsp;positions &lt;- sample(seq_len(ncol(sequences)), size=nMutations)<br><br>
    
                &emsp;&emsp;&emsp;&emsp;# Create the mutations<br>
                &emsp;&emsp;&emsp;&emsp;sequences[row, positions] &lt;- <br>
                &emsp;&emsp;&emsp;&emsp;&emsp;&emsp;sample(c("A", "C", "G", "T"), replace=TRUE, <br>
                &emsp;&emsp;&emsp;&emsp;&emsp;&emsp;size=nMutations)<br><br>
                &emsp;&emsp;}<br><br>
  
                &emsp;&emsp;return(sequences)<br>
                }<br>
            </div>

            <p>
                With these functions we can create a random alignment like this:
            </p>

            <div id="codeBlock">

                # Set the number of sequences<br>
                nSequences &lt;- 25<br><br>

                # Set the number of sites in the sequences<br><br>
                nSites &lt;- 200<br><br>

                # Create a random nucleotide alignment<br>
                sequences &lt;- createNucleotideSequenceAlignment(nSequences, nSites, meanNSNPs=10)<br>
            </div>

            <p>
                Finally we get to use my <c>plotFASTA()</c> function:
            </p>    

            <div id="codeBlock">
                plotFASTA(sequences)
            </div>

            <img src="plotFASTA_Thumbnail.png" width="95%" hspace="0.5%" vspace="1%">

            <p>
                Hopefully it provides a handy way you can take a quick look at the nucleotide alignment in your FASTA file to check it looks ok. It is very similar to the output from the <c>image.DNAbin()</c> function, but it is less picky about the format of the nucleotide alignment.<br><br>


                If you're interested, you can see how <c>plotFASTA()</c> function works <a href="https://github.com/JosephCrispell/basicPlotteR/blob/master/R/plotFasta.R">here</a> on github. Enjoy, and I hope it is useful!! :-)<br>
            </p>  

        </div>

    </body>

</html>
