
<html>

    <head>
        <title>Joseph Crispell | Blog - tanglePlot</title>
        <link href="../../default.css" rel="stylesheet" type="text/css">

        <!-- Global site tag (gtag.js) - Google Analytics -->
        <script async src="https://www.googletagmanager.com/gtag/js?id=UA-109829558-1"></script>
        <script>
            window.dataLayer = window.dataLayer || [];
            function gtag(){dataLayer.push(arguments);}
            gtag('js', new Date());
            gtag('config', 'UA-109829558-1');
        </script>
    </head>

    <body>

        <!-- Insert Sidebar box -->
        <div id="sidebar" style="top: 43.5%">

            <pageLink><a href="../../index.html">Home</a></pageLink>
            <pageLink><a href="../../Blog.html">Blog</a></pageLink>
            <pageLink><a href="../../About.html">About</a></pageLink>
            <pageLink><a href="../../Outputs.html">Outputs</a></pageLink>
        </div>

        <!-- Insert Title -->
        <div id="title">
            <img src="../../BlogThumbnail.png" width="40%">

        </div>

        <!-- Insert Text Body -->
        <div id="mainText">

            <br><br><br>

            <h1> Comparing phylogenies with tanglePlot()&emsp;08-12-19</h1>

            <p>
                It's cold and wet outside and I've a wee bit of time spare to write a post about my <c>tanglePlot()</c> function.<br><br>

                Because there are so many ways to build a phylogeny from a nucleotide alignment (for example by using neighbour joining, maximum likelihood, or Bayesian algorithms), we can often end up with slightly different phylogenies.<br><br>

                For a quick comparison, you can use <c>tanglePlot()</c>. Here's how it works:
            </p>

            <div id="codeBlock">

              # Set the seed<br>
              set.seed(76263)<br><br>

              # Create a random phylogeny<br>
              tree &lt;- ape::rtree(n=50, rooted=TRUE)<br><br>

              # Rotate one of the nodes in the phylogeny - with tip labels it<br>
              rotated &lt;- ape::rotate(tree, node=c("t35", "t34"))<br><br>

              # Compare the random and rotated phylogeny<br>
              tanglePlot(tree, rotated, connectingLine.col="red", connectingLine.lty=2,<br>
              &emsp;&emsp;&emsp;&emsp;show.tip.label=TRUE, offsetProp=0.02)
            </div>

            <img src="tanglePlot_Thumbnail.png" width="95%" hspace="0.5%" vspace="1%">

            <p>
                The <c>tanglePlot()</c> plots two phylogenies opposite one another and adds lines to show where tips are in different places. Here, I created a random phylogeny and rotated one of it's nodes to illustrate how <c>tanglePlot()</c> works.<br><br>

                Hopefully you'll find this useful. I've added <c>tanglePlot()</c> to a new R package I am building (<c>phyloHelpeR</c>). You can install <c>phyloHelpeR</c> using:
            </p>

            <div id="codeBlock">
              install.packages("devtools")<br>
              devtools::install_github("JosephCrispell/phyloHelpeR")<br>
            </div>

            <p>
                I am still building the <c>phyloHelpeR</c> package so let me know if you have any problems!<br><br>

                If you're interested, the code for <c>tanglePlot()</c> is available <a href="https://github.com/JosephCrispell/phyloHelpeR/blob/master/R/tanglePlot.R">here</a>. There are a couple of very interesting bits of code:
            </p>

            <div id="codeBlock">
              # Get the coordinates of the tips on the current plotted phylogeny<br>
              lastPP &lt;- get("last_plot.phylo", envir = .PlotPhyloEnv)<br><br>

              # Convert between relative and absolute coordinates in plotting window<br>
              absolute &lt;- grconvertX(relativeCoords, "user", "ndc")
            </div>

        </div>

    </body>

</html>
